x-pack/packages/kbn-elastic-assistant/impl/assistant/conversations/conversation_settings/use_conversation_changed.tsx

```diff
@@ -0,0 +1,105 @@
+import { ConversationsBulkActions } from '../../api';
+import { AIConnector } from '../../../connectorland/connector_selector';
+
+interface Props {
+  allSystemPrompts: Prompt[];
+  conversationSettings: Record<string, Conversation>;
+  conversationsSettingsBulkActions: ConversationsBulkActions;
+  defaultConnector?: AIConnector;
+  setConversationSettings: React.Dispatch<React.SetStateAction<Record<string, Conversation>>>;
+  setConversationsSettingsBulkActions: React.Dispatch<
+    React.SetStateAction<ConversationsBulkActions>
+  >;
+  onSelectedConversationChange: (conversation?: Conversation) => void;
+}
+
+type OnConversationSelectionChange = (c?: string | Conversation) => void;
+
+export const useConversationChanged = ({
+  allSystemPrompts,
+  conversationSettings,
+  conversationsSettingsBulkActions,
+  defaultConnector,
+  setConversationSettings,
+  setConversationsSettingsBulkActions,
+  onSelectedConversationChange,
+}: Props) => {
+  const defaultSystemPrompt = useMemo(() => {
+    return getDefaultSystemPrompt({ allSystemPrompts, conversation: undefined });
+  }, [allSystemPrompts]);
+
+  // Conversation callbacks
+  // When top level conversation selection changes
+  const onConversationSelectionChange: OnConversationSelectionChange = useCallback(
+    (c = '') => {
+      const isNew = typeof c === 'string';
+
+      const newSelectedConversation: Conversation | undefined = isNew
+        ? {
+            id: '',
+            title: c ?? '',
+            category: 'assistant',
+            messages: [],
+            replacements: {},
+            ...(defaultConnector
+              ? {
+                  apiConfig: {
+                    connectorId: defaultConnector.id,
+                    actionTypeId: defaultConnector.actionTypeId,
+                    provider: defaultConnector.apiProvider,
+                    defaultSystemPromptId: defaultSystemPrompt?.id,
+                  },
+                }
+              : {}),
+          }
+        : c;
+
+      if (newSelectedConversation && (isNew || newSelectedConversation.id === '')) {
+        setConversationSettings({
+          ...conversationSettings,
+          [isNew ? c : newSelectedConversation.title]: newSelectedConversation,
+        });
+        setConversationsSettingsBulkActions({
+          ...conversationsSettingsBulkActions,
+          create: {
+            ...(conversationsSettingsBulkActions.create ?? {}),
+            [newSelectedConversation.title]: newSelectedConversation,
+          },
+        });
+      } else if (newSelectedConversation != null) {
+        setConversationSettings((prev) => {
+          return {
+            ...prev,
+            [newSelectedConversation.id]: newSelectedConversation,
+          };
+        });
+      }
+
+      onSelectedConversationChange({
+        ...newSelectedConversation,
+        id: newSelectedConversation.id || newSelectedConversation.title,
+      });
```

Comment:
id can be title sometimes.

Reply:
https://github.com/elastic/kibana/assets/6295984/589c4297-9e5c-4c3e-830d-fa8167b8335d

Reply:
@YulNaumenko, thoughts on persisting all the base conversations on initial conversation request that way we can get rid of all this extra title vs id logic? I don't like side-effects on a GET, but can't think of a different way of handling this than what we're doing now, which has been a source of bugs requiring conditional checks like this.